transports/websys/src/websocket.rs

```diff
@@ -0,0 +1,321 @@
+                // generates error on the browser console we just signal it to the
+                // stream.
+                shared.lock().error = true;
+            }
+        });
+        socket.set_onerror(Some(error_callback.as_ref().unchecked_ref()));
+
+        let close_callback = Closure::<dyn FnMut(_)>::new({
+            let shared = shared.clone();
+            move |_| {
+                let mut locked = shared.lock();
+                locked.closed = true;
+                if let Some(waker) = &locked.waker {
+                    waker.wake_by_ref();
+                }
+            }
+        });
+        socket.set_onclose(Some(close_callback.as_ref().unchecked_ref()));
+
+        // Manage closures memory.
+        let closures = SendWrapper::new((
+            open_callback,
+            message_callback,
+            error_callback,
+            close_callback,
+        ));
+
+        shared.lock().closures = Some(closures);
+
+        Self { shared }
+    }
+}
+
+impl AsyncRead for Connection {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let mut shared = self.shared.lock();
+        shared.waker = Some(cx.waker().clone());
+
+        if shared.error {
+            Poll::Ready(Err(io::Error::new(io::ErrorKind::Other, "Socket error")))
+        } else if shared.closed {
+            Poll::Ready(Err(io::ErrorKind::BrokenPipe.into()))
+        } else if shared.data.is_empty() {
+            Poll::Pending
+        } else {
+            let n = shared.data.len().min(buf.len());
+            for k in buf.iter_mut().take(n) {
+                *k = shared.data.pop_front().unwrap();
+            }
+            Poll::Ready(Ok(n))
+        }
+    }
+}
+
+impl AsyncWrite for Connection {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        let mut shared = self.shared.lock();
+        shared.waker = Some(cx.waker().clone());
+
+        if shared.error {
+            Poll::Ready(Err(io::Error::new(io::ErrorKind::Other, "Socket error")))
+        } else if shared.closed {
+            Poll::Ready(Err(io::ErrorKind::BrokenPipe.into()))
+        } else if !shared.opened {
+            Poll::Pending
+        } else {
+            match shared.socket.send_with_u8_array(buf) {
+                Ok(()) => Poll::Ready(Ok(buf.len())),
+                Err(err) => Poll::Ready(Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    format!("Write error: {err:?}"),
+                ))),
+            }
+        }
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, _: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        Poll::Pending
```

Comment:
We need to return `Poll::Ready` eventually here, otherwise we will never be able to close these streams. I think we should:

- Check if we are closed, if yes return `Poll::Ready`
- If not and we are also not `closing`, call `Websocket::close` and set a `closing` flag to true
- If we are `closing`, return `Poll::Pending` and set a waker
- Once we are called in `on_close`, call the waker